pwiz_tools/Skyline/Util/LazyComboBox.cs

```diff
@@ -0,0 +1,141 @@
+﻿/*
+ * Original author: Brian Pratt <bspratt .at. proteinms dot net>,
+ *                  MacCoss Lab, Department of Genome Sciences, UW
+ *
+ * Copyright 2021 University of Washington - Seattle, WA
+ * 
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+using System;
+using System.Collections.Generic;
+using System.Drawing;
+using System.Windows.Forms;
+using pwiz.Skyline.Properties;
+
+namespace pwiz.Skyline.Util
+{
+    /// <summary>
+    /// Substitute for ComboBox which avoids performance issues by only
+    /// creating its dropdown list when user action demands it
+    /// </summary>
+    public class LazyComboBox : Button
+    {
+        public List<object> Items;
+        private ListBox _listbox;
+        private int _selectedIndex;
+
+        public event EventHandler SelectedIndexChanged;
+
+        public int SelectedIndex
+        {
+            get => _selectedIndex;
+            set
+            {
+                var oldValue = _selectedIndex;
+                _selectedIndex = value;
+                base.Text = _selectedIndex < 0 ? string.Empty : Items[_selectedIndex].ToString();
+                if (value != oldValue)
+                {
+                    SelectedIndexChanged?.Invoke(this, EventArgs.Empty);
+                }
+            }
+        }
+
+        public override string Text
+        {
+            get => base.Text;
+            set => SelectedIndex = FindStringExact(value);
+        }
+
+        public LazyComboBox()
+        {
+            Items = new List<object>();
+            base.TextAlign = ContentAlignment.MiddleLeft;
+            FlatStyle = FlatStyle.Flat;
+            SelectedIndex = -1;
+            AutoEllipsis = true; // e.g. If test is too wide, show "Explicit Coll..." rather than breaking at space and showing "Explicit"
+            base.BackColor = SystemColors.ControlLight;
+            FlatAppearance.BorderSize = Math.Max(1, FlatAppearance.BorderSize/2); // Allow text closer to edge than in a normal button, like standard ComboBox does
+            // Show a dropdown hint
+            Image = Resources.DropImageNoBackground;
+            ImageAlign = ContentAlignment.MiddleRight;
+            TextImageRelation = TextImageRelation.Overlay;
+        }
+
+        public int FindStringExact(string str)
+        {
+            return Items.IndexOf(str);
+        }
+
+        protected override void OnClick(EventArgs e)
+        {
+            if (_listbox == null)
```

Comment:
I worry a bit about this ListBox use where the listbox is actually parented inside the form, but apparently above all other controls. I would have used a menu, since they are designed for this. If the form is not tall enough, this implementation seems like it would clip the listbox.

Reply:
I'll try a menu, thanks.